daml-script/test/daml/upgrades/UpgradesTest.daml

```diff
@@ -0,0 +1,594 @@
+assertUnknownPackageError (Right _) = assertFail $ "Expected missing package error, but submission succeeded"
+
+-- Convenient wrapper we can inline with the submit call
+liftAssertUnknownPackageError : Script (Either SubmitError a) -> Script ()
+liftAssertUnknownPackageError s = s >>= assertUnknownPackageError
+
+-- Any queries or submits run within this block will have the package-id present.
+withExplicitPackageIds : Script a -> Script a
+withExplicitPackageIds act = do
+  setProvidePackageId True
+  ret <- act
+  setProvidePackageId False
+  pure ret
+
+main : Script ()
+main = forA_ tests $ \(testName, test) -> do
+  debugRaw $ "Testing: " <> testName
+  test
+
+tests : [(Text, Script ())]
+tests = mconcat
+  [ -- "Single" participant tests
+    -- (we only submit/read from one participant, but by nature of having multiple participants on the same domain, version selection is affected)
+    [ -- daml3-script query tests
+      ("Upgrade a contract when querying", queryUpgraded)
+    , ("Downgrade a contract with Nones when querying", queryDowngradedNone)
+    , ("Fail to downgrade a contract with Somes when querying", queryDowngradedSome)
+
+    , -- Fetching tests
+      ("Upgrade a contract when fetching", fetchUpgraded)
+    , ("Downgrade a contract with Nones when fetching", fetchDowngradedNone)
+    , ("Fail to downgrade a contract with Somes when fetching", fetchDowngradedSome)
+
+    , -- Fetching tests with unvetted sources (BROKEN)
+      ("(BROKEN) Upgrade a contract when fetching where the source package (V1) is unvetted", fetchUpgradedSourceUnvetted)
+    , ("(BROKEN) Downgrade a contract with Nones when fetching where the source package (V2) is unvetted", fetchDowngradedNoneSourceUnvetted)
+
+    , -- Template payload upgrading/downgrading from exercises within choice bodies
+      -- The expected behaviour here is that the outer templates have fixed the choice version they wish to call
+      -- So the ledger must coerce the real data to match the expected data. Regardless of what the "most recent" version of that package is
+      ("Call a V1 choice on a V1 contract from within a choice body, expect V1 implementation is used regardless of the possible upgrade.", exerciseV1ChoiceV1ContractSameType)
+    , ("Call a V1 choice on a V2 contract from within a choice body, expect V1 implementation is used, via a downgrade.", exerciseV1ChoiceV2ContractSameType)
+    , ("Call a V1 choice on a V2 contract with Some field from within a choice body, expect V1 implementation to attempt to be used, but fail downgrade", exerciseV1ChoiceV2ContractSameTypeSome)
+    , ("Call a V2 choice on a V1 contract from within a choice body, expect V2 implementation is used, via an upgrade", exerciseV2ChoiceV1ContractSameType)
+    , ("Call a V2 choice on a V2 contract from within a choice body, expect V2 implementation is used, for sanity", exerciseV2ChoiceV2ContractSameType)
+
+    , -- Choice upgrading
```

Comment:
Another sanity test would be to call a V1 contract with a V2 choice that doesn't exist on the V1 contract.

Reply:
Good shout, hadn't considered new choices